automation/taskcluster/lib/tasks.py

```diff
@@ -2,143 +2,517 @@
+        self, task_id, repo_url, branch, commit, owner, source, scheduler_id, date_string,
+        tasks_priority='lowest'
+    ):
         self.task_id = task_id
+        self.repo_url = repo_url
+        self.branch = branch
+        self.commit = commit
         self.owner = owner
         self.source = source
         self.scheduler_id = scheduler_id
-        self.build_worker_type = build_worker_type
+        self.tasks_priority = tasks_priority
+        self.date = arrow.get(date_string)
 
-    def build_task(self, name, description, command, artifacts, features, scopes=[], routes=[]):
-        created = datetime.datetime.now()
-        expires = taskcluster.fromNow('1 year')
-        deadline = taskcluster.fromNow('1 day')
+    def craft_build_task(self, apks, is_staging=False):
+        artifacts = {
+            'public/{}'.format(os.path.basename(apk)): {
+                "type": 'file',
+                "path": apk,
+                "expires": taskcluster.stringDate(taskcluster.fromNow(DEFAULT_EXPIRES_IN)),
+            }
+            for apk in apks
+        }
 
-        return {
-            "workerType": self.build_worker_type,
-            "taskGroupId": self.task_id,
-            "schedulerId": self.scheduler_id,
-            "expires": taskcluster.stringDate(expires),
-            "retries": 5,
-            "created": taskcluster.stringDate(created),
-            "tags": {},
-            "priority": "lowest",
-            "deadline": taskcluster.stringDate(deadline),
-            "dependencies": [self.task_id],
-            "routes": routes,
-            "scopes": scopes,
-            "requires": "all-completed",
-            "payload": {
-                "features": features,
-                "maxRunTime": 7200,
-                "image": "mozillamobile/fenix:1.3",
-                "command": [
-                    "/bin/bash",
-                    "--login",
-                    "-c",
-                    command
-                ],
-                "artifacts": artifacts,
-                "deadline": taskcluster.stringDate(deadline)
+        sentry_secret = '{}project/mobile/fenix/sentry'.format(
+            'garbage/staging/' if is_staging else ''
+        )
+        leanplum_secret = '{}project/mobile/fenix/leanplum'.format(
+            'garbage/staging/' if is_staging else ''
+        )
+
+        pre_gradle_commands = (
+            'python automation/taskcluster/helper/get-secret.py -s {} -k {} -f {}'.format(
+                secret, key, target_file
+            )
+            for secret, key, target_file in (
+                (sentry_secret, 'dsn', '.sentry_token'),
+                (leanplum_secret, 'production', '.leanplum_token'),
+            )
+        )
+
+        gradle_commands = (
+            './gradlew --no-daemon -PcrashReports=true clean test assembleRelease',
+        )
+
+        command = ' && '.join(
+            cmd
+            for commands in (pre_gradle_commands, gradle_commands)
```

Comment:
I put the `pre_gradle` one separate because it has some logic in there. I don't want to confuse future readers when they see a list-comprehension inside a list.